src/main/scala/com/microsoft/hyperspace/index/dataskipping/DataSkippingIndex.scala

```diff
@@ -0,0 +1,174 @@
+      deletedSourceDataFiles: Seq[FileInfo],
+      indexContent: Content): (Index, Index.UpdateMode) = {
+    if (appendedSourceData.nonEmpty) {
+      writeImpl(ctx, index(ctx, appendedSourceData.get), SaveMode.Overwrite)
+    }
+    if (deletedSourceDataFiles.nonEmpty) {
+      val spark = ctx.spark
+      import spark.implicits._
+      val deletedFileIds = deletedSourceDataFiles.map(_.id).toDF(IndexConstants.DATA_FILE_NAME_ID)
+      val updatedIndexData = spark.read
+        .parquet(indexContent.files.map(_.toString): _*)
+        .join(deletedFileIds, Seq(IndexConstants.DATA_FILE_NAME_ID), "left_anti")
+      val writeMode = if (appendedSourceData.nonEmpty) {
+        SaveMode.Append
+      } else {
+        SaveMode.Overwrite
+      }
+      writeImpl(ctx, updatedIndexData, writeMode)
+    }
+    val updateMode = if (deletedSourceDataFiles.isEmpty) {
+      Index.UpdateMode.Merge
+    } else {
+      Index.UpdateMode.Overwrite
+    }
+    (this, updateMode)
+  }
+
+  override def refreshFull(
+      ctx: IndexerContext,
+      sourceData: DataFrame): (DataSkippingIndex, DataFrame) = {
+    val updatedIndex = copy(sketches = ExpressionUtils.resolve(ctx.spark, sketches, sourceData))
+    (updatedIndex, updatedIndex.index(ctx, sourceData))
+  }
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case DataSkippingIndex(thatSketches, _) => sketches.toSet == thatSketches.toSet
+      case _ => false
+    }
+
+  override def hashCode: Int = sketches.map(_.hashCode).sum
+
+  /**
+   * Creates index data for the given source data.
+   */
+  def index(ctx: IndexerContext, sourceData: DataFrame): DataFrame = {
+    val fileNameCol = "input_file_name"
+    val indexDataWithFileName = sourceData
+      .groupBy(input_file_name().as(fileNameCol))
+      .agg(aggregateFunctions.head, aggregateFunctions.tail: _*)
+
+    // Construct a dataframe to convert file names to file ids.
+    val spark = ctx.spark
+    val relation = RelationUtils.getRelation(spark, sourceData.queryExecution.optimizedPlan)
+    import spark.implicits._
+    val fileIdDf = ctx.fileIdTracker
+      .getIdToFileMapping(relation.pathNormalizer)
+      .toDF(IndexConstants.DATA_FILE_NAME_ID, fileNameCol)
+
+    indexDataWithFileName
+      .join(
+        fileIdDf.hint("broadcast"),
+        IndexUtils.decodeInputFileName(indexDataWithFileName(fileNameCol)) ===
+          fileIdDf(fileNameCol))
+      .select(
+        IndexConstants.DATA_FILE_NAME_ID,
+        indexDataWithFileName.columns.filterNot(_ == fileNameCol).map(c => s"`$c`"): _*)
+  }
+
+  private def writeImpl(ctx: IndexerContext, indexData: DataFrame, writeMode: SaveMode): Unit = {
+    indexData.cache()
+    val minRowCountPerFileConfig =
+      HyperspaceConf.DataSkipping.minRecordsPerIndexDataFile(ctx.spark)
+    val maxNumFiles = toIntExact(math.max(1, indexData.count() / minRowCountPerFileConfig))
+    val repartitionedIndexData = indexData.coalesce(maxNumFiles)
```

Comment:
Ok one question: the index data - is it already partitioned? Why are you worry about small files? 
If it's **max**RecordPerFiles, we don't need to care about it.

I think coalesce is not a good idea. it just control the number of files, not the size or number of record for each files. 
And the performance is sometimes worse than repartition.

Reply:
Because often data skipping index data is very small and Spark is bad at working with many small files. Without repartition, the tiny index data is stored in too many files like 200 files. It's no good for performance.

I think coalesce is fine - it doesn't increase the number of files and I've verified that it evenly distributes rows over files.

Reply:
Because of the join, it's partitioned. coalesce merges partitions, not distributing the rows as it's optimized for that purpose. 
So.. as you said performance doesn't matter for small index dataset, I would prefer `repartition(maxNumFiles)`

Reply:
Okay, then we need to bring back the branch because it shouldn't increase the number of files.

Reply:
I think it's okay to repartition with numPartitions based on maxRowCountPerFiles config as users can adjust it and the default value(100000) looks reasonable. `indexData.count()`  (+ for minRowCountPerFileData groupby/count) incurs spark jobs & caching operation even it's faster enough.
So I would avoid it if possible. cc @imback82 WDYT?

This is my preference:
```scala
    val maxRowCountPerFileConfig =
      HyperspaceConf.DataSkipping.maxRecordsPerIndexDataFile(ctx.spark)
    val maxNumFiles = toIntExact(math.max(1, indexData.inputFiles.length / maxRowCountPerFileConfig))
    val repartitionedIndexData = indexData.repartition(maxNumFiles)
```

Reply:
The whole point of the code here is to optimize the index application time. Index creation/refresh times are also important, but we should not reverse the priority. If it helps the application time, it makes sense to pay the small cost during creation/refresh. Index creation/refresh are maintenance operations, and the code for cache, groupBy, count, all combined, takes only a fraction of the total creation/refresh time. Most of the time is spent during reading the source data and computing sketch aggregation functions.

Reply:
@clee704 I agree with your point, but what I'm trying to say is your approach and suggested approach do the *same* thing, but yours has more additional operations that I'd like to avoid.

Reply:
No, it's not the same, as the suggested approach fixes the number of files, which can be larger than it was before repartition.

Reply:
Anyway, this row count based repartitioning thing was only an approximation for a better approach which is based on the data size. I'll push a new commit.

Reply:
With the default config (100,000) and the shuffle partition config (200), it only happens when it has more than (200  *  100,000) files. Considering smaller shuffle partition number, let say 1, numFiles would be between 1~10 for 100K~1M files. 

And if a user wants to reduce the index file size, maybe because of large BFs, they tries a smaller config val, but it won't work because of the condition.

Reply:
That's why I said it was a suboptimal approach. For the best index application performance, index data files should be evenly distributed over files in terms of data size because index data filtering is mostly I/O bound.